src/Compression/CompressionCodecSZ3.cpp

```diff
@@ -0,0 +1,285 @@
+    /// Therefore, update SZ3 with care to avoid breaking existing persistencies.
+    /// We mark it as experimental for now.
+    bool isExperimental() const override { return true; }
+    bool isVectorCodec() const override { return true; }
+    String getDescription() const override { return "SZ3 is a lossy compressor for floating-point data with error bounds."; }
+
+private:
+    UInt32 doCompressData(const char * source, UInt32 source_size, char * dest) const override;
+    void doDecompressData(const char * source, UInt32 source_size, char * dest, UInt32 uncompressed_size) const override;
+
+    UInt32 getMaxCompressedDataSize(UInt32 uncompressed_size) const override;
+
+    std::vector<size_t> dimensions;
+    const UInt8 float_width;
+    const SZ3::ALGO algorithm;
+    const SZ3::EB error_bound_mode;
+    const Float64 error_value;
+};
+
+namespace ErrorCodes
+{
+    extern const int BAD_ARGUMENTS;
+    extern const int CORRUPTED_DATA;
+    extern const int ILLEGAL_CODEC_PARAMETER;
+    extern const int LOGICAL_ERROR;
+}
+
+CompressionCodecSZ3::CompressionCodecSZ3(
+    UInt8 float_size_, SZ3::ALGO algorithm_, SZ3::EB error_bound_mode_, double error_value_)
+    : float_width(float_size_)
+    , algorithm(algorithm_)
+    , error_bound_mode(error_bound_mode_)
+    , error_value(error_value_)
+{
+    setCodecDescription("SZ3");
+}
+
+uint8_t CompressionCodecSZ3::getMethodByte() const
+{
+    return static_cast<uint8_t>(CompressionMethodByte::SZ3);
+}
+
+void CompressionCodecSZ3::updateHash(SipHash & hash) const
+{
+    getCodecDesc()->updateTreeHash(hash, true);
+    hash.update(float_width);
+}
+
+UInt32 CompressionCodecSZ3::getMaxCompressedDataSize(UInt32 uncompressed_size) const
+{
+    return uncompressed_size + sizeof(UInt8);
+}
+
+UInt32 CompressionCodecSZ3::doCompressData(const char * source, UInt32 source_size, char * dest) const
+{
+    SZ3::Config config;
+
+    std::vector<size_t> result_dimensions;
+    size_t floats = source_size / float_width;
+    for (auto dimension : dimensions)
+        floats /= dimension;
+
+    result_dimensions.push_back(floats);
+    for (auto dimension : dimensions)
+        result_dimensions.push_back(dimension);
+
+    config.setDims(result_dimensions.begin(), result_dimensions.end());
+
+    config.cmprAlgo = algorithm;
+    config.errorBoundMode = error_bound_mode;
+
+    switch (error_bound_mode)
+    {
+        case SZ3::EB_REL:
+            config.relErrorBound = error_value;
+            break;
+        case SZ3::EB_ABS:
+            config.absErrorBound = error_value;
+            break;
+        case SZ3::EB_PSNR:
+            config.psnrErrorBound = error_value;
+            break;
+        case SZ3::EB_L2NORM:
+            config.l2normErrorBound = error_value;
+            break;
+        default:
+            throw Exception(ErrorCodes::LOGICAL_ERROR, "Invalid error bound mode");
+    }
+
+    char * compressed;
```

Comment:
This together with `delete[]` in l. 165 looks like a potential memleak (we should not trust the sz3 API to behave sane). It would be cool to wrap the array in a unique_ptr for automatic cleanup (RAII).